### DISCUSSION TITLE
Plugin.cfg strange description of ordering

### DISCUSSION THREAD

**Discussion by renbou** [coredns/coredns#4800] [Q&A] [answered]
What is meant by "Every plugin will feel the effects of all other plugin below (after) them during a request" in plugin cfg? How is it that a plugin will feel the effects of plugins below them? If I'm correct, the plugins will be called in top-to-bottom order, so how is it that, for example, plugin "metadata" which is above plugin "file" will feel the effects of the "file" plugin? Are plugins called in such fashion that after the return of the next plugin in the chain, the previous, "parent", plugin will get the return value or something like that?

**Comment by chrisohaver** [comment, marked as answer]
I agree, I also find the wording here confusing.

I have kind of assumed that the writer had in mind the plugins executing in a call stack, in which case the order of before and after might be reversed from a list if the stack is imagined as piling upwards.

The reality of how a request is affect by plugins is more complex though and depends on the plugin.  Plugins _are_ executed in a call stack. Each plugin calls the next before returning. So each plugin has the opportunity to act on a DNS request when pushed on the stack _and_ when popped off the stack. Most plugins however just take action when being pushed on the stack.

**Reply by chrisohaver** [reply]
> Each plugin calls the next before returning.

To clarify, the entire list of plugins is not traversed for every request.  If a plugin has an answer for a request, then it _doesn't_ call the next plugin before returning.

**Reply by renbou** [reply]
Ah, ok, that's what I've guessed, but for me it sounded strange to have a plugin do something after the request has been handled by some later plugin. Maybe it makes sense for some logging plugins though, sure. Thank you!

**Reply by renbou** [reply]
> > Each plugin calls the next before returning.
> 
> To clarify, the entire list of plugins is not traversed for every request. If a plugin has an answer for a request, then it _doesn't_ call the next plugin before returning.

Makes sense, that's why the order of plugins matters in plugin.cfg
